### ISSUE TITLE
Message passing base class doubts

### ISSUE THREAD

**Issue by yorard123** [rusty1s/pytorch_geometric#350]
Hi,
I have some small doubts that I would like to solve. I will appreciate if you can help me to clarify.

1. I would like to know which criterion you used to put the linear transformation of the GCNConv implementation of the tutorial in the forward pass and in the NNConv implementation. In contrast, a similar transformation is implemented inside the message method. 

```
class GCNConv(MessagePassing):
    def __init__(self, in_channels, out_channels):
        super(GCNConv, self).__init__(aggr='add')  # "Add" aggregation.
        self.lin = torch.nn.Linear(in_channels, out_channels)

    def forward(self, x, edge_index):
        # x has shape [N, in_channels]
        # edge_index has shape [2, E]

        # Step 1: Add self-loops to the adjacency matrix.
        edge_index, _ = add_self_loops(edge_index, num_nodes=x.size(0))

        # Step 2: Linearly transform node feature matrix.
        **x = self.lin(x)**

        # Step 3-5: Start propagating messages.
        return self.propagate(edge_index, size=(x.size(0), x.size(0)), x=x)

    def message(self, x_j, edge_index, size):
       ...
    def update(self, aggr_out):
      ...
```
```

class NNConv(MessagePassing):
   def __init__(self,
                 in_channels,
                 out_channels,
                 nn,
                 aggr='add',
                 root_weight=True,
                 bias=True):
...

    def forward(self, x, edge_index, edge_attr):
        """"""
        x = x.unsqueeze(-1) if x.dim() == 1 else x
        pseudo = edge_attr.unsqueeze(-1) if edge_attr.dim() == 1 else edge_attr
        return self.propagate(edge_index, x=x, pseudo=pseudo)

    def message(self, x_j, pseudo):
        **weight = self.nn(pseudo).view(-1, self.in_channels, self.out_channels)**
        return torch.matmul(x_j.unsqueeze(1), weight).squeeze(1)

    def update(self, aggr_out, x):
     ...
```

If in NNConv for any reason  you need to apply to the output of the transformation a regularization layer that it needs to be applied in a neighborhood level (I'm speculating, I am only trying to understand properly the design decisions in order to be consistent in my code), will you put this transformation in the forward method or in the message method? 

2. I have a doubt with the nomenclature. In the message passing base class when the flow is source to target the x_i corresponds to source and x_j to target, Am I right? Another way to see it is x_i indicates the node and x_j its neighbors, right?

3. Related to the last question. I am curious to see the implementation of this "sub-variables" I mean you convert the variable x to x_i and x_j, this logic where is it implemented? I would like to see how is implemented, I like so much this feature of this framework =)

4. Just to confirm, the message method supports any number of variables that I wish to use, right? Need I consider any limitation on this?

**Comment by rusty1s** [comment]
Hi,
1. Well, the linear layer of `GCNConv` will transform all nodes the same, while the MLP of the `NNConv` applies a transformation on the edge attributes (or pseudo-coordinates). Therefore, we need to call the MLP of the `NNConv` in the `message` function. In contrast, we could also call the linear transformation of `GCNConv` in the `message` function, but this would result in a lot of unnecessary computations. Therefore, it is more efficient to transform the features before propagating messages.
2. No. `x_i` and `x_j` will have the same meaning, regardless of the flow you use: `x_j` corresponds to the neighboring node features that get aggregated, while `x_i` denotes the features of the central node that is in the process of aggregation. This way you can change the flow without changing the underlying code. In the default case of `flow='source_to_target'`, edges are denoted as `(j,i)`, and if `flow='target_to_source'`, edges are denoted as `(i,j)`.
3. You can find the implementation [here](https://github.com/rusty1s/pytorch_geometric/blob/master/torch_geometric/nn/conv/message_passing.py), though it might be hard to grasp due to its abstractness.
4. Exactly: No limitations.

**Comment by yorard123** [comment]
Hi,

> Well, the linear layer of GCNConv will transform all nodes the same, while the MLP of the NNConv applies a transformation on the edge attributes (or pseudo-coordinates). Therefore, we need to call the MLP of the NNConv in the message function. In contrast, we could also call the linear transformation of GCNConv in the message function, but this would result in a lot of unnecessary computations. Therefore, it is more efficient to transform the features before propagating messages.

I see. Related to this I have a doubt on the message method. The message method only process a pair of nodes the central and one neighboring node, isn' it? Then if I need to apply to the output of the mlp an activation layer that needs to be applied in the neighborhood. In this case, the mlp will be located on the forward method, right?

**Comment by rusty1s** [comment]
I am sorry, I do not know what you mean with "applied in the neighborhood". Can you give more details?

**Comment by yorard123** [comment]
Yes, sorry I think it is not very well explained. You have the neighborhood N that contains edges E. These edges are passed through the mlp `output_neigh = mlp(E)` and to this output (you want to apply the activation layer `output_neigh = activation(output_neigh)` Is it more clear?

**Comment by rusty1s** [comment]
Well, you are free to use an activation layer on each message or on each aggregated output (this is not equivalent). So you can either do:
```python
def forward(self, x, edge_index, pseudo):
    out = self.propagate(edge_index, x=x, pseudo=pseudo)

def message(self, pseudo, x_j):
    return activation(self.mlp(pseudo) * x_j)
```
or
```python
def forward(self, x, edge_index, pseudo):
    out = self.propagate(edge_index, x=x, pseudo=pseudo)
    return activation(out)

def message(self, pseudo, x_j):
    return self.mlp(pseudo) * x_j
```

**Comment by yorard123** [comment]
Ok, then the message is called for every pair of connected nodes, right? Then if I put, for instance, a softmax I will have different behavior depending on if I used in the message method or in the forward method.  Right?

**Comment by rusty1s** [comment]
Technically, the `message` function gets only called once, and processes each pair of connected nodes  in parallel.

I now understand your question better due to you mentioning softmax. So you want to normalize your features across each neighborhood. This is exactly what we do in the [`GATConv`](https://github.com/rusty1s/pytorch_geometric/blob/master/torch_geometric/nn/conv/gat_conv.py#L95). This needs to be done pre-aggregation in the `message` function. You need to use the softmax function provided by PyG which works across neighbors. Normalization should be done via `edge_index_i`, which group all node pairs based on the *same* central node.

**Comment by yorard123** [comment]
This `edge_index_i` it will be the same as if I use the `src` variable obtained from `src, dst = edge_index`?

**Comment by yorard123** [comment]
Because I do not want to apply the softmax to the node features, I want to apply this softmax to the edge_attr. But the problem that I have is.. the gpu mem requirements increase a lot...

**Comment by rusty1s** [comment]
Ah I see what you mean. Of course, you can call the MLP on the edge attributes before `propagate`. I was actually referring to the additional matrix multiplication with the node features (which needs to be performed in `message`). Efficiency is equivalent.

Concerning your memory consumptions: I never noticed this in practice, but, of course, the softmax function has an additional memory footprint, which could potentially be decreased by implementing a custom sparse softmax kernel.

**Comment by yorard123** [comment]
> Concerning your memory consumptions: I never noticed this in practice

I suppose that it is only noticed in big graphs or deep networks.

> the softmax function has an additional memory footprint, which could potentially be decreased by implementing a custom sparse softmax kernel.

Is it difficult to do it? I am not experienced in cuda but I suppose it would be great to have a optimized softmax function.

**Comment by rusty1s** [comment]
I will add it to my todo.

**Comment by yorard123** [comment]
Thanks!
For my side we can close the issue. Thank you very much, for your help!
